src/object.d

```diff
@@ -41,6 +41,7 @@ version (D_ObjectiveC) public import core.attribute : selector;
 public import rt.array.comparison : __cmp;
 public import rt.array.equality : __ArrayEq, __equals;
 public import rt.array.casting: __ArrayCast;
+public import rt.array.concat : _d_arraycatnTX, _d_arraycatnTXTrace;
```

Comment:
I don't like that we have these public implementation specific functions. Can the compiler look directly in `rt.array.concat`? Or can it be a private import and have the compiler bypass the protection?

Reply:
These templates are instantiated by user code, not by the compiler.  The compiler just lowers expressions to them.  So they need to be, source code and all, available to user programs.

Reply:
Yes, I understand that. But currently all these functions are a weird hybrid of public API and compiler internal API. 

Either these functions should be called by the user and in that case they should be well documented and have good names following the standard naming conventions. Or they should only be called by the compiler. In the latter case we should try to make them only be callable by the compiler and not by users.

Reply:
They can't be *called* by the compiler.  They can *only* be instantiated, and then called by user code.  However, the user should never instantiate and call these implementations explicitly.  The user writes `auto concatenatedArray = array1 ~ array2;`  The compiler lowers that to something like `auto concatenatedArray = _d_arraycatnTX([array1, array2]);` specifically so the user doesn't have to instantiate and call the template explicitly.  Perhaps I'm not understanding.  I intend to add a upcoming PR to do something like this...

Reply:
For example: `_d_arraycatnTX` can be private and the compiler can lower the code to:

```d
auto concatenatedArray = __traits(getMember, rt.array, "_d_arraycatnTX")([array1, array2]);
```

Because `__traits(getMember)` can bypass the protection. Although I just noticed that it's possible to access a private function using `__traits(getMember)` but it's not possible to call the private function. I'm not sure if that's intentional.

Reply:
Well, the compiler could disable access checks for these specific calls.

However, I'd simply keep things as-is for now and later on change the compiler to directly access the hooks in the original modules, not in object.d: `import rt.array : _d_arraycatnTX = __some_name_which_will_never_conflict`. This change might be somewhat controversial, so it's better to do this later on and not stall further progress.